### ISSUE TITLE
with .. do .. else .. end

### ISSUE THREAD

**Issue by benwilson512** [elixir-lang/elixir#4085]
I want to look at the viability of allowing an else clause on the newly introduced `with`. Right now `with` obviously flattens the success cases of matching, but it also effectively flattens the error cases. Unfortunately there isn't a built in mechanism with `with` to take advantage of the flattened error cases. So for example a common pattern that has arisen in my code around building dependent ecto models is:

```elixir
  def create(params) do
    with {:ok, organization} <- create_organization(params),
         {:ok, address}      <- create_address(organization, params) do
      {:ok, %{organization | address: address}}
    end
    |> case do
      {:ok, organization} -> {:ok, organization}
      {:error, changeset} -> {:error, changeset |> error_string}
    end
  end
```

This could more succinctly written if `with` had an else clause

```elixir
  def create(params) do
    with {:ok, organization} <- create_organization(params),
         {:ok, address}      <- create_address(organization, params) do
      {:ok, %{organization | address: address}}
    else
      {:error, changeset} -> {:error, changeset |> error_string}
    end
  end
```

Thoughts?

**Comment by lexmag** [comment]
I've had this discussion with @fishcakez on IRC, as I remember we agreed that it'd be nice to have `else` which works in a similar way to `else` in `try`. (James, please correct me if I'm wrong.)

**Comment by alco** [comment]
@lexmag Could you clarify? `else` in `try` is evaluated if the expression did not fail. According to @benwilson512's proposal, `else` in `with` should be invoked if a match fails. Otherwise it won't add anything beyond what can be achieved by piping `with` into `case do`.

**Comment by lexmag** [comment]
@alco I mean `else` in `with` should be similar to `else` in `try` in term of supporting patterns and not to be just any expression like `else` in `if`, invocation of course is different.
In other words, it is what is proposed here.

**Comment by benwilson512** [comment]
I'm glad you brought up the use case of `else` in `try` as I was worried that using `else` with matches lacked precedent.

I suppose the remaining question is what to do if all of the `else` clauses fail. `MatchError` ? That seems the right choice to me. If you use `with` without `else` you match on success cases and return any error cases. If you use `with` as well as `else` you're exhaustively enumerating both success and failure patterns, or running into a MatchError just like `case`

**Comment by alco** [comment]
`try` raises a `TryClauseError`.

A small correction to @benwilson512's comment:
> If you use `with` as well as `else` you're exhaustively enumerating both success and failure patterns, [...]

You're not enumerating successes, only failures.

I think your proposal can be described in simpler terms:

Using `with`, one can write a sequence of matches and if all of them succeed, the body between `do` and `end` will only be evaluated. If a match fails, no further expressions are evaluated and the non-matched value is returned. If there is an `else` clause, the non-matched value will be passed to it where it will be matched against a list of clauses like in a `case`. This allows us to adjust what is being returned from `with` in the case of a failed match.

**Comment by almassapargali** [comment]
Totally would love to see this in Elixir, makes a lot of sense, and already come up with some cases I'd use this.

**Comment by almassapargali** [comment]
Not going further, if @benwilson512 used `Repo.transaction`, his code could be:
```elixir
    Repo.transaction fn ->
      with {:ok, organization} <- create_organization(params),
           {:ok, address}      <- create_address(organization, params) do
        %{organization | address: address}
      else
        {:error, changeset} -> Repo.rollback(changeset)
      end
    end
```

Of course we can do this with `|> case do`, but I think this one is much cleaner.

**Comment by alco** [comment]
;)
```elixir
Repo.transact_with {:ok, organization} <- create_organization(params),
                   {:ok, address}      <- create_address(organization, params)
do
  %{organization | address: address}
else
  # This point is reached after the transaction rolls back
  {:error, changeset} -> Repo.rollback(changeset)
end
```

---

offtopic: I wish Elixir supported line breaking with `\` or similar, to be able to write code like this:
```elixir
Repo.transact_with \
  {:ok, organization} <- create_organization(params),
  {:ok, address}      <- create_address(organization, params)
do
...
end
```

Function-name-length-independent indentation helps a lot when reading the code IMO.

**Comment by josevalim** [comment]
> offtopic: I wish Elixir supported line breaking with \ or similar, to be able to write code like this:

It **does**. :)

**Comment by alco** [comment]
Awesome! Somehow I thought it wasn't supported when I tried it a few days ago.

**Comment by vic** [comment]
Just for reference, I created a tiny [happy](http://github.com/vic/happy) macro that just converts code into a chain of cases, I'm currently using it in a couple of projects of mine, as IMHO it's cleaner than `with`'s syntax. It also allows an `else` clause for handling non-matching values. The downside is variables would _leak_ as it's just a plain old case.

**Comment by liveforeverx** [comment]
@vic For variables not leaking in your implementation, you can transform expression in happy_path macro to with expression instead of cases.

I personally find `with` cleaner as `happy_path`, because it is much more explicit in a case, where to define cases and where not:

```elixir
with {:ok, width} <- Map.fetch(opts, :width),
     double_width = width * 2,
```

And, after using it our codebase, later separation of do I find better, because of better separation of concerns. And most use cases, where I used were like:

```elixir
with .... <- get_data
     .... <- another_get_or_validating
do
   pure_transformation
end
```

where `happy_path` will be

```elixir
happy_path do
 .... = get_data
 .... = another_get_or_validating
 pure_transformation
end
```

But it is really, my IMHO.

**Comment by pdxrod** [comment]
Can anyone tell me what %{organization | address: address} does?

**Comment by fertapric** [comment]
@pdxrod from the Elixir guides https://elixir-lang.org/getting-started/keywords-and-maps.html

> Maps have the following syntax for updating a key’s value:
> ```elixir
> iex> map = %{:a => 1, 2 => :b}
> %{2 => :b, :a => 1}
> 
> iex> %{map | 2 => "two"}
> %{2 => "two", :a => 1}
> iex> %{map | :c => 3}
> ** (KeyError) key :c not found in: %{2 => :b, :a => 1}
> ```

However, I would recommend you to use Elixir Forum or Elixir's Slack for such questions. The issue tracker should be reserve for issues :)
